### PR TITLE
[PM-6072] Add get$ method on SendService

### PR DESCRIPTION
**Pull request by aj-rosado**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Adding a reactive method to return a Send observable

## Code changes

- **send.service.ts:** Creates a new method named get$ that returns an observable which observes a single value from the send$ observable.
- **send.service.spec.ts:** Added tests to validate the `get$` returns the expected values

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/tools/send/services/send.service.abstraction.ts [hunk 1]
```diff
@@ -18,7 +18,16 @@ export abstract class SendService {
...
     key?: SymmetricCryptoKey,
   ) => Promise<[Send, EncArrayBuffer]>;
+  /**
+   * @deprecated Do not call this, use the get$ method
+   */
   get: (id: string) => Send;
+  /**
+   * Provides a send for a determined ID
+   * @param id The id of the desired send
+   * @returns An observable that listens to the value of the desired send
```
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.abstraction.ts:28]: 👍🏻 Documentation!!!!!! Thank you!
🎨 I've been including when the observable updates in the interface documentation.

File: libs/common/src/tools/send/services/send.service.spec.ts [hunk 2]
```diff
@@ -67,6 +67,20 @@ describe("SendService", () => {
     });
   });
 
+  describe("get$", () => {
```
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.spec.ts:70]: ⚠️ This should also test an update of the observable.

File: libs/common/src/tools/send/services/send.service.ts [hunk 3]
```diff
@@ -116,6 +116,10 @@ export class SendService implements InternalSendServiceAbstraction {
     return sends.find((send) => send.id === id);
   }
 
+  get$(id: string): Observable<Send | undefined> {
+    return this.sends$.pipe(map((sends) => sends.find((o) => o.id === id)));
```
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.ts:120]: 🤔 This will transmit update every time _any_ send updates. Is there a way it could be focused to only update when the matching send updates? If we can solve that problem here, then it could be reused for the `SendStateProvider`.

**Review by audrey-jensen — CHANGES_REQUESTED** [review]
👍 for great changes
🤔 for open inquiry that could benefit from discussion
🎨 for suggestions 
⚠️ for concerns needing attention

"request changes" applies only to the ⚠️. The rest are just comments and observations.

File: libs/common/src/tools/send/services/send.service.ts (hunk 3 above)
Comment (aj-rosado) [reply libs/common/src/tools/send/services/send.service.ts:120]: As the new observable returned by `get$` only is observing a single Send object, I would expect it to only transmit update when that object itself is updated.

Is not the case here?

Comment (audrey-jensen) [reply libs/common/src/tools/send/services/send.service.ts:120]: Unfortunately, no. The observer [emits a new value every time `sends$` emits](https://stackblitz.com/edit/l86urm?file=index.ts). In order to get the observable to only update when there's a change, you'll need to run it through something like [`distinctUntilChanged`](https://rxjs.dev/api/operators/distinctUntilChanged).

Comment (aj-rosado) [reply libs/common/src/tools/send/services/send.service.ts]: Added the `distinctUntilChanged` operator, I've run various tests seems to be working as expected, only calling subscribe when that Send has a change in value.

I still want to run some more tests, principally on possible null cases.

File: libs/common/src/tools/send/services/send.service.abstraction.ts (hunk 1 above)
Comment (aj-rosado) [reply libs/common/src/tools/send/services/send.service.abstraction.ts:29]: Added

File: libs/common/src/tools/send/services/send.service.spec.ts (hunk 2 above)
Comment (aj-rosado) [reply libs/common/src/tools/send/services/send.service.spec.ts:70]: Added tests to validate if it is updating as expected and if it emits on subscribe the expected amount of times

File: libs/common/src/tools/send/services/send.service.ts (hunk 3 above)
Comment (audrey-jensen) [reply libs/common/src/tools/send/services/send.service.ts]: There's also quite a few lines that need code coverage. You might be able to simply the tests if you extract the function and test it separately.

**Review by audrey-jensen — COMMENTED** [review]
The changes I asked for are all addressed. I'll approve once the Codecov comments are addressed. If there's a certain one you want to leave in, please explain why w/ a comment in the PR. :)

File: libs/common/src/tools/send/services/send.service.spec.ts [hunk 4]
```diff
@@ -67,6 +71,123 @@ describe("SendService", () => {
...
+
+      await sendService.replace({
+        "1": sendData("1", "Test Send Updated"),
+      });
+
+      const result2 = await firstValueFrom(singleSendObservable);
+      expect(result2).toEqual(send("1", "Test Send Updated"));
+    });
+
+    it("subscribe counter", async () => {
```
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.spec.ts:100]: ❓ I don't understand the behavior being tested here. Are you trying to test that each of the branches in the `distinctUntilChanged` comparator are followed?

🎨 If so, this should be broken down into separate tests, with each `it` labelled according to the behavior being tested.

```typescript
it("reports a change when name changes on a new send", () => {
   let changed = false;
   sendService.get$("1).subscribe(() => {
      changed = true;
   });
   const sendDatObject = new SendData({} as any);
   sendDataObject.id = "1";

   await sendService.replace({ 
      "1": sendDataObject,
      "2": sendData("2", "Test Send 2");
   });

  expect(changed).toEqual(true);
});
```

When the tests are atomic in this fashion, the behavior being tested is very clear. This makes determining the causes of failure pretty straightforward. As written, if any of the `replace` calls fails, then you need to debug the test to see which didn't hit your `subCounter++` line.

🎨 Something else that can help make these tests straightforward is making a helper function to configure the object:

```typescript
const defaultSendData = {
  id: "1",
  name: "Test Send",
  disabled: false,
  // ... and so on
};

createSendData(value: Partial<SendData> = {}) {
    const testSend = new SendData({} as any);
    for(let prop in defaultSendData) {
        testSend[prop] = value[prop] ?? defaultSendData[prop];
    }
}

// Now you can create the object you'd like to test easily
const sendDataObject = createSendData();

// and if you need a specialized value for a test
const sendDataObject = createSendData({ expirationDate: "2024-04-10" });
```

File: libs/common/src/tools/send/services/send.service.ts [hunk 5]
```diff
@@ -116,6 +116,68 @@ export class SendService implements InternalSendServiceAbstraction {
...
+      distinctUntilChanged((oldSends, newSends) => {
+        const oldSend = oldSends.find((oldSend) => oldSend.id === id);
+        const newSend = newSends.find((newSend) => newSend.id === id);
+        if (!oldSend || !newSend) {
+          // If either oldSend or newSend is not found, consider them different
+          return false;
+        }
+
+        // Compare each property of the old and new Send objects
+        const allPropertiesSame = Object.keys(newSend).every((key) => {
```
Comment (audrey-jensen) [inline libs/common/src/tools/send/services/send.service.ts:130]: 🤔 This loop seems more complex than it should be. Checking whether any key switched from a `null` value to a `not null` value makes a lot of sense in the loop. I wonder, though, whether the switch might be more clear with a series of more specific checks.

```typescript
// if a key is missing on one send but not the other, they're different
const keys = Object.keys(newSend) as (keyof Send)[];
if(keys.find((key) => oldSend[key] && !newSend[key])) {
  return true;
}

// if an encrypted field is different, they're different
const encryptedFields: (keyof Send)[] = ["name", "notes", "key"];
if(encryptedFields.find(f => oldSend[f].encryptedString !== newSend[f].encryptedString) {
  return true;
}
encryptedFields.push("text");
if(oldSend.text.text?.encryptedString === newSend.text.text?.encryptedString) {
  return true;
}

// if a time field is different, they're different
const timeFields: (keyof Send)[] = ["revisionDate", "expirationDate", "deletionDate"];
if(timeFields.find(f => oldSend[f]?.getTime() !== newSend[f]?.getTime())) {
  return true;
}

// fields that never change can be ignored
const ignoredFields: (keyof Send)[] = ["file"];

// if any other field is different, they're different
const checkedFields: (keyof Send)[] = [...encryptedFields, ...timeFields, ...ignoredFields];
const remainingFields = keys.filter(f => !checkedFields.includes(f));
if(remainingFields.find((f) => oldSend[f] !== newSend[f]) {
   return true;
}

// they're the same
return false;
```

I really don't know. This is probably eye of the beholder, and need not be addressed.

**Review by audrey-jensen — COMMENTED** [review]
I'm 90% sure the unit tests need to be factored a little differently.

File: libs/common/src/tools/send/services/send.service.spec.ts (hunk 4 above)
Comment (aj-rosado) [reply libs/common/src/tools/send/services/send.service.spec.ts:100]: At first this was just a test to verify if the subscribe was only called when the intended Send was changed.
While founding some edge cases and covering the whole case it grew a lot more than I initially expected and it now verifies if each case from `distinctUntilChanged` is working properly.

Agree it needs to be reworked and thanks for the suggestions

**Review by audrey-jensen — APPROVED** [review]
